### DISCUSSION TITLE
How does one debug a textual app?

### DISCUSSION THREAD

**Discussion by hucker** [Textualize/textual#835] [Q&A]
I am running on windows.  I've found that if I run python 3.8, I can get the samples to run in power shell and cmd.  If I run from pycharm, I get a bunch of '.' characters in the "terminal", if I run in vscode, I get a nicely rendered screen, but there is no mouse interaction, so testing anything that isn't write-only isn't very useful.  For example, if you run the calculator app VSCode's windows terminal screen displays a nice calculator, but you can't click on the buttons and get a response

I was going to try to make changes to the calculator (because real calculators are RPN) but quickly found debugging problematic since vscode/pycharm don't seem to detect mouse events when the app is running.  I'd like to be able to run an app, set a breakpoint in a mouse handler, and debug from there "normally."  Can such debugging in IDEs , or does one need to drop down to PDB or use remote debugging methods?

**Comment by jeffwright13** [comment]
I've not had the trouble of no-mouse in VSCode when debugging, but I am running on a Mac. Dumb question, did you click into the Terminal window when you hit your breakpoint? 

In general, async apps are harder to debug than standard apps, but Textual does provide some nice traceback information (prettily formatted with Rich) and if you scroll back in the terminal, you can usually see where things went wrong.

**Comment by zzzeek** [comment]
just in case other newbs are here searching for "how to debug", a very basic facility appears to be the [console](https://textual.textualize.io/guide/devtools/#console)

**Comment by Toniolo-Marco** [comment]
I had the same problem. The solution is to edit your pycharm run/debug configuration: `Modify options ->  Emulate terminal in output console` check this option. It is not 100% accurate with the colors or focus effects, and mouse scroll doesn't work, you have to drag the bar; but it does the trick. Notice that I just use `python` command, not `textual console` or anything
